src/libraries/System.IO.Compression.Zstandard/src/System/IO/Compression/ZstandardEncoder.cs

```diff
@@ -0,0 +1,487 @@
+        /// <param name="bytesWritten">The number of bytes written to the destination.</param>
+        /// <param name="isFinalBlock">True if this is the final block of data to compress.</param>
+        /// <returns>An <see cref="OperationStatus"/> indicating the result of the operation.</returns>
+        /// <exception cref="ObjectDisposedException">The encoder has been disposed.</exception>
+        public OperationStatus Compress(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten, bool isFinalBlock)
+        {
+            EnsureNotDisposed();
+
+            bytesConsumed = 0;
+            bytesWritten = 0;
+
+            if (source.IsEmpty && !isFinalBlock)
+                return OperationStatus.Done;
+
+            return CompressCore(source, destination, out bytesConsumed, out bytesWritten,
+                isFinalBlock ? Interop.Zstd.ZstdEndDirective.ZSTD_e_end : Interop.Zstd.ZstdEndDirective.ZSTD_e_continue);
+        }
+
+        /// <summary>Flushes any remaining processed data to the destination buffer.</summary>
+        /// <param name="destination">The buffer to write the flushed data to.</param>
+        /// <param name="bytesWritten">The number of bytes written to the destination.</param>
+        /// <returns>An <see cref="OperationStatus"/> indicating the result of the operation.</returns>
+        /// <exception cref="ObjectDisposedException">The encoder has been disposed.</exception>
+        public OperationStatus Flush(Span<byte> destination, out int bytesWritten)
+        {
+            EnsureNotDisposed();
+
+            return CompressCore(ReadOnlySpan<byte>.Empty, destination, out _, out bytesWritten,
+                Interop.Zstd.ZstdEndDirective.ZSTD_e_flush);
+        }
+
+        private OperationStatus CompressCore(ReadOnlySpan<byte> source, Span<byte> destination,
+            out int bytesConsumed, out int bytesWritten, Interop.Zstd.ZstdEndDirective endDirective)
+        {
+            bytesConsumed = 0;
+            bytesWritten = 0;
+
+            unsafe
+            {
+                fixed (byte* inBytes = &MemoryMarshal.GetReference(source))
+                fixed (byte* outBytes = &MemoryMarshal.GetReference(destination))
+                {
+                    var input = new Interop.Zstd.ZstdInBuffer
+                    {
+                        src = (IntPtr)inBytes,
+                        size = (nuint)source.Length,
+                        pos = 0
+                    };
+
+                    var output = new Interop.Zstd.ZstdOutBuffer
+                    {
+                        dst = (IntPtr)outBytes,
+                        size = (nuint)destination.Length,
+                        pos = 0
+                    };
+
+                    nuint result = Interop.Zstd.ZSTD_compressStream2(_context!, ref output, ref input, endDirective);
+                    if (ZstandardUtils.IsError(result))
```

Comment:
When would this be reached?
`srcSize_wrong` reads like a caller (our) error, i.e. it shouldn't be reachable.